packages/grid/x-data-grid/src/hooks/features/filter/gridFilterUtils.ts

```diff
@@ -234,20 +236,60 @@ export const buildAggregatedFilterItemsApplier = (
     return null;
   }
 
-  return (row, shouldApplyFilter) => {
-    const resultPerItemId: GridFilterItemResult = {};
-
-    for (let i = 0; i < appliers.length; i += 1) {
-      const applier = appliers[i];
-      if (!shouldApplyFilter || shouldApplyFilter(applier.item.field)) {
-        resultPerItemId[applier.item.id!] = applier.v7
-          ? applier.fn(row)
-          : applier.fn(getRowId ? getRowId(row) : row.id);
-      }
-    }
-
-    return resultPerItemId;
-  };
+  // Original logic:
+  // return (row, shouldApplyFilter) => {
+  //   const resultPerItemId: GridFilterItemResult = {};
+  //
+  //   for (let i = 0; i < appliers.length; i += 1) {
+  //     const applier = appliers[i];
+  //     if (!shouldApplyFilter || shouldApplyFilter(applier.item.field)) {
+  //       resultPerItemId[applier.item.id!] = applier.v7
+  //         ? applier.fn(row)
+  //         : applier.fn(getRowId ? getRowId(row) : row.id);
+  //     }
+  //   }
+  //
+  //   return resultPerItemId;
+  // };
+
+  // We generate a new function with `eval()` to avoid expensive patterns for JS engines
+  // such as a dynamic object assignment, e.g. `{ [dynamicKey]: value }`.
+  const filterItemTemplate = `(function filterItem$$(row, shouldApplyFilter) {
+      ${appliers
+        .map(
+          (applier, i) =>
+            `const shouldApply${i} = !shouldApplyFilter || shouldApplyFilter(${JSON.stringify(
+              applier.item.field,
+            )});`,
+        )
+        .join('\n')}
+
+      const result$$ = {
+      ${appliers
+        .map(
+          (applier, i) =>
+            `${applier.item.id !== undefined ? parseInt(String(applier.item.id), 10) : 'undefined'}:
+          !shouldApply${i} ?
+            false :
+            ${
+              applier.v7
+                ? `appliers[${i}].fn(row)`
+                : `appliers[${i}].fn(${getRowId ? 'getRowId(row)' : 'row.id'})`
+            },
+      `,
+        )
+        .join('\n')}};
+
+      return result$$;
+    })`;
+
+  // eslint-disable-next-line no-eval
+  const filterItem = eval(
+    filterItemTemplate.replaceAll('$$', String(filterItemsApplierId)),
```

Comment:
I wonder what makes the function unique? Wasn't a unique function name enough to solve https://github.com/mui/mui-x/pull/9635/files#r1258665866?

Reply:
No I think it's the function args + body. Essentially the content of the [`new Function(...)` constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/Function).